### ISSUE TITLE
Make RSS feed distinguish between author, title, date, and abstract in data

### ISSUE THREAD

**Issue by keithgee** [VTUL/vtechworks#384]
We need more information before we can work on this task, @amandafrench.

Are you trying to integrate with another system? Would you be able to use the atom feed?
http://vtechworks.lib.vt.edu/feed/atom_1.0/site

```
<entry>
<title>Vegetative Potential to Reduce Total Dissolved Solid Nutrient Ions Generated from Reclaimed Mine Lands in Central Appalachia</title>
<link href="http://hdl.handle.net/10919/72873" rel="alternate"/>
<author>
<name>Gondran, Amy Christine</name>
</author>
<id>http://hdl.handle.net/10919/72873</id>
<updated>2016-09-04T07:07:43Z</updated>
<published>2016-09-02T00:00:00Z</published>
<summary type="text">Vegetative Potential to Reduce Total Dissolved Solid Nutrient Ions Generated from Reclaimed Mine Lands in Central Appalachia
Gondran, Amy Christine
One of the major issues in surface coal mine land reclamation is the impact of total dissolved solids (TDS) on water quality. To address this issue, this study assessed whether vegetation could reduce TDS nutrient ion movement from the rooting zone in early reclamation stages when TDS generation is often highest. Vegetated and un-vegetated paired plots were established across eight sites with gradients in age, spoil material, and vegetative productivity. Ion exchange resin lysimeters were used to compare nutrient ion fluxes in soil solution that contribute to TDS between paired plots. Soil and vegetation properties (used as proxies for evapotranspiration and plant uptake) were characterized at each site and correlated with log ratios of common TDS contributing ions [i.e., base cations Ca2+, Mg2+, Na+, K+ (RBC), sulfate (RSO42-), and total Ca2+, Mg2+, Na+, K+, SO42- (Rtotal)] between un-vegetated and vegetated plots. Strong Spearman correlations were found between RTotal, RBC, and RSO42- during the peak growing season, and were weakened overall during vegetative dormancy. Soil organic matter was shown to be a strong correlate through dormant periods. Correlations between TDS nutrient ions and vegetation suggest evapotranspiration in the growing season, and interception during dormant periods, exert an influence on nutrient ion fluxes. These findings indicate that aggrading forests may reduce TDS nutrient ion loading, through solute retention mechanisms driven by organic matter inputs and plant uptake.
</summary>
<dc:date>2016-09-02T00:00:00Z</dc:date>
</entry>
```

The atom feed already has entries for author (author/name), title (title), date (published), and abstract (summary), with the problem that summary also repeats the title and author. Are you asking for the author and title to be removed from the summary data from the atom feed?

Please clarify. And if you let us know the purpose of this task, we might be able to accommodate it through other means.
Thanks!

**Comment by amandafrench** [comment]
Hmm, that code looks fine, but if you go to any of the RSS feed links, e.g., http://vtechworks.lib.vt.edu/feed/atom_1.0/site, Title, ID (handle), Summary, and Date are the only elements displayed. The Author field data isn't displayed.

![screenshot 2016-09-09 13 13 02](https://cloud.githubusercontent.com/assets/227453/18395977/2e51419c-768f-11e6-84ae-e9b96a712c6f.png)

Also, since the Summary element contains the title, author, and abstract in one element, that's all run together in the RSS display. It's clear, too, that Title, ID (handle), Summary, and Date are the only elements being used by other applications "reading" the RSS -- see these screenshots of the RSS feed embedded in WordPress and the source code for it: 

![screenshot 2016-09-09 13 16 13](https://cloud.githubusercontent.com/assets/227453/18396053/9c7df944-768f-11e6-88fc-f2f65719a70b.png)

![screenshot 2016-09-09 13 17 25](https://cloud.githubusercontent.com/assets/227453/18396093/c9241f8c-768f-11e6-8175-5e079575a09d.png)

Basically, this task, if it's possible, has two parts: to get the `<author>` element to display in the RSS feed and to remove the `<summary>` element from the RSS feed. 

Sorry wasn't clearer!

**Comment by amandafrench** [comment]
Oops, and a third part: get the abstract alone into the RSS feed if possible, without the title and author as it is in the `<summary>` field.

**Comment by amandafrench** [comment]
And yes: the purpose of this task is to make sure that all RSS feeds from VTechWorks contain and deliver appropriate structured data: linked Title, Author, Date, and if possible an excerpt from the abstract so that this data can be used by other applications (RSS readers, subscription clients, websites).

**Comment by keithgee** [comment]
Thanks - I did have to "View Source" on the Atom feed to see the author data.

**Comment by pgray64** [comment]
RSS feed options are in the main dspace config. Dspace intentionally only inserts the author field into the Atom feed. Coincidentally, the Atom feed is the one that does not validate with W3C, which might be why the author field is not showing despite the author tags looking valid in the source. 

Additionally W3C validator reports "Use of unknown namespace: http://apache.org/cocoon/i18n/2.1" for all 3 feeds.

I can modify java source files to make Dspace insert author fields into RSS 1.0 and 2.0 as well.

**Comment by pgray64** [comment]
The RSS 1.0 and 2.0 specifications don't actually support multiple authors and the author field is parsed as an email. That explains why author was omitted by DSpace for RSS 1.0 and 2.0. 

@amandafrench Do you still want me to remove the authors from the summary field, since for RSS 1.0 and 2.0 there is no other way to display authors? I'll keep on researching why the Atom feed author field doesn't actually render in any feed readers, despite being present in the source.

**Comment by pgray64** [comment]
DSpace uses a prebuilt jar file (rome-1.0.jar) for the actual rendering of the feeds. So there is no way to specify options beyond those available in the main DSpace config file. The java source files that can be modified really only pass the parameters in the main config file to the classes in the prebuilt rome-1.0.jar.

**Comment by amandafrench** [comment]
If I've followed all that, @pgray64, then it should be possible to display the linked title, the date, and the author in all three feeds if you remove the title and the abstract/description from the summary field by editing the java source files. Is that right? If so, I think that's worth doing. If the only place the authors can display is in the summary field, then let's use that field for author(s). Feel free to hit me up on chat or come to my office next week if I've misunderstood or if there are further questions you need answered right away.

**Comment by alawvt** [comment]
@amandafrench, rome-1.0.jar is a prebuilt byte code package which is called by DSpace. The [Rome project](https://rometools.github.io/rome/index.html) is documented at https://rometools.github.io/rome/TutorialsAndArticles.html. @pgray64 might be able to research this project to determine whether he can call it's methods, etc. to get the data the way that you want it. (My quick estimate is that it would be fairly involved.) Would you like him to research whether that would be possible and if so, how long that might take?

**Comment by amandafrench** [comment]
Hi @alawvt and @pgray64 -- the short answer is yes, although clearly this isn't doable by the end of this sprint. The longer answer is that I'd like @pgray64 to figure out some method of fixing this not just for us but for anyone who's running our same version of DSpace, for instance Texas A&M: https://oaktrust.library.tamu.edu/feed/rss_2.0/site It is actually pretty high priority because it relates to publicizing the material in VTechWorks; if we can get a legible RSS feed whose individual fields can be styled with CSS, then I can confidently tell departments that they can embed it on their department sites. If that involves adding more RSS configuration options to dspace.cfg or some other high-level thing, so be it. 

Sigh. DSpace!!! /me shakes fist 

I'll leave the issue open, because I anticipate that it'll make it into the next sprint plan, but no need to work on it anymore this sprint, @pgray64

**Comment by keithgee** [comment]
Philippe has done research and started working on this issue and I hope to learn from him how the RSS feeds work soon.

**Comment by pgray64** [comment]
I'm currently checking if using ROME 1.7 fixes the fact that current RSS feeds are not passing validators and how author fields for atom feeds are not getting rendered.

**Comment by pgray64** [comment]
ROME outputs a document object that can be traversed and modified before being returned to the browser. Since ROME lacks options for things like adding authors to rss 1.0/2.0 and adding custom CSS files, I'll write some custom code to add those things before returning the feed. I'll make them configurable options as well.

Its kind of hacky to be modifying the XML document after its rendered by ROME, but I cannot find any ROME options in their Javadoc or documentation that lets us do what we need.

**Comment by pgray64** [comment]
@amandafrench when you mentioned CSS styling did you want the feeds to be styled with our own custom stylesheet or did you mean that departments who embed the feeds on their websites should be able to specify their own stylesheet?

**Comment by amandafrench** [comment]
@pgray64 Thanks for working on this, @pgray64. I meant both, really, though also neither. I just want a particular kind of data to be confined to a particular field (element), and if that's the case, then anyone could in theory style that element with CSS. Example: 

```
<item>
<title>Characterization of the Arabidopsis glutamine dumper1 mutant reveals connections between amino acid homeostasis and plant stress responses</title>
<link>http://hdl.handle.net/10919/73175</link>
<description>Characterization of the Arabidopsis glutamine dumper1 mutant reveals connections between amino acid homeostasis and plant stress responses
Yu, Shi
Amino acids constitute the major organic form of transported nitrogen in plants, elements for protein synthesis, and precursors of many plant secondary metabolites, such as lignin, hormones, and flavonoids. Furthermore, amino acid metabolism lies at the crossroad of carbon and nitrogen metabolism. The Arabidopsis glutamine dumper1 (gdu1) mutant secretes glutamine from hydathodes, a phenotype caused by the overexpression of Glutamine Dumper1 (GDU1). GDU1 is a small transmembrane protein presents only in higher plants. The gdu1-1D mutant shows a pleiotropic phenotype: perturbed amino acid metabolism, tolerance to exogenous toxic concentrations of amino acids, elevated amino acid export, and activated stress/defense responses, lesions, and smaller rosettes. The biochemical function of GDU1 remains elusive. To better elucidate the biological processes leading to the complex Gdu1D phenotype, two approaches were conducted: (1) An ethyl methanesulfonate suppressor screening of the Gdu1D phenotype, which led to the isolation of intragenic mutations in GDU1 and mutations in the ubiquitin ligase LOG2 (Loss Of Gdu1D 2). Study of the intragenic mutations in GDU1 helped to characterize its structure-function relationships. Characterization of LOG2 showed that LOG2 interacts with GDU1 and is necessary for the Gdu1D phenotype. (2) The responses of the plant to the dexamethasone-induced expression of GDU1 were studied over time. This experiment identified major signaling pathways contributing to different components of the Gdu1D phenotype and the early events triggered by the perturbation of amino acid homeostasis. Our results showed that GDU1 overexpression first increases amino acid export, which is followed by amino acid imbalance and stress responses. This study sheds light on how amino acid imbalance interacts with various plant signaling pathways and stress responses, and suggests that LOG2 is involved in this process.
</description>
<pubDate>Wed, 15 Apr 2015 00:00:00 GMT</pubDate>
<guid isPermaLink="false">http://hdl.handle.net/10919/73175</guid>
<dc:date>2015-04-15T00:00:00Z</dc:date>
</item>
```

In that example, because the title, author, and abstract all appear in the `<description>` element, I can't style the description author so that it appears on a separate line or in a different font or whatever. In that example, I could use CSS something like 

```
item title {font-size: x-large;} 
```

but anything I do with `item description` would affect everything in that description field, not the author and title and abstract separately. 

(I don't claim that the above is valid CSS, but you get my drift.) 

CSS styling isn't that important, though. I just want 1) the author to show up in as many feeds and browsers as practicable, and 2) the `<description>` to have just the abstract and not ALSO the author and title. Or as close to that as possible.

**Comment by pgray64** [comment]
Okay, I see. I'll leave in the CSS option changes I made because it will make testing the other (more important) requirements easier - Firefox seems to not render RSS feeds without an inbuilt style.

Working on a possible fix that puts authors in a comma separated field for RSS 1.0 and 2.0, since those two don't allow multiple author fields.

**Comment by amandafrench** [comment]
Thanks, @pgray64. If it can only show the first author that's fine by me. Remember too that I think this should be done not so much just for us as for the DSpace project as a whole; the community might have input into how best to do this. Actually of course they probably already agreed on doing it the way it is, but I hate that way; looks terrible when displayed, which is the whole point. 

Let me know if you aren't able to access the test page on the R&I development website at http://dev.researchinformatics.vtlibraries.net/rss-test/. You and @keithgee should both have gotten login credentials by email.

**Comment by amandafrench** [comment]
@pgray64 And I would be very happy to have CSS options to help style the RSS feed, so thanks!

**Comment by pgray64** [comment]
My new approach is to write a ROME "module generator" which is essentially a plugin to ROME that adds extra fields to items before RSS feeds are generated. Much cleaner an upgrade-safe than the DOM manipulation I was trying before.

**Comment by pgray64** [comment]
Found a much easier fix. It turns out that a simple change in dspace.cfg makes all three RSS feeds have dc.creator fields, one field for each author, which is exactly what we want except we want the tag to be called "author" instead of "dc.creator"
So the existing code for the DCModule can be reused to work for author fields too.

**Comment by keithgee** [comment]
👍🏽@pgray64 great!

**Comment by amandafrench** [comment]
Really?!? Awesome!! Thanks @pgray64!

**Comment by amandafrench** [comment]
And actually the tag can be called "creator" instead of "author" as far as I'm concerned -- I don't know if that matters as long as the value itself shows up. If it does matter, then by all means do whatever makes sense.

**Comment by pgray64** [comment]
Pull request #424

Changes:
`dc.creator` tags contain each author for RSS 1.0 and 2.0 and `author` tag is still used for Atom.
This is actually considered good practice because the author tag is supposed to only be for email addresses for RSS 1.0 and RSS 2.0 specs. 

Whether or not `dc.creator` or `author` is rendered is dependent on what client you use. There is no standard as to what fields RSS viewers actually render and it seems that almost all by default do not display either `author` or `dc.creator`. I tried a bunch of clients with other websites' valid feeds with `author` and `dc.creator` and those fields did not display. 

I'm fairly certain that is why DSpace originally listed authors in the description.

**Comment by amandafrench** [comment]
@pgray64 Drat, I knew there was probably a good reason. For me it's still a problem in ANY client that the display looks garbled because the `<description>` repeats the title and runs it together with the author(s) and the abstract. Let me look at the code and see what it looks like in the RSS & Atom feeds when built. Thanks.

**Comment by alawvt** [comment]
Resolved with PR #424

**Comment by amandafrench** [comment]
@pgray64 Apparently the RSS weirdness is an existing bug with the XMLUI version of DSpace: https://jira.duraspace.org/browse/DS-1543 -- in the JSPUI version everything is lumped into the ```<description>``` field too, but there are text labels such as "Authors:" prepended, which makes it more legible.
